packages/plugins/hmac-upstream-signature/src/index.ts

```
@@ -0,0 +1,136 @@
+  let fetchAPI: FetchAPI;
+  let textEncoder: TextEncoder;
+
+  return {
+    onYogaInit({ yoga }) {
+      fetchAPI = yoga.fetchAPI;
+    },
+    onSubgraphExecute({ subgraphName, subgraph, executionRequest, setExecutionRequest }) {
+      if (shouldSign({ subgraphName, subgraph, executionRequest })) {
+        textEncoder ||= new fetchAPI.TextEncoder();
+        key$ ||= createCryptoKey({
+          textEncoder,
+          crypto: fetchAPI.crypto,
+          secret: options.secret,
+          usages: ['sign'],
+        });
+        return key$.then(async key => {
+          const serializedExecutionRequest = serializeExecutionRequest(executionRequest);
+          const encodedContent = textEncoder.encode(serializedExecutionRequest);
+          const signature = await fetchAPI.crypto.subtle.sign('HMAC', key, encodedContent);
+          const extensionValue = btoa(String.fromCharCode(...new Uint8Array(signature)));
+          setExecutionRequest({
+            ...executionRequest,
+            extensions: {
+              ...executionRequest.extensions,
+              [extensionName]: extensionValue,
+            },
+          });
+        });
+      }
+    },
+  };
+}
+
+export type HMACUpstreamSignatureValidationOptions = {
+  secret: string;
+  extensionName?: string;
+  serializeParams?: (params: GraphQLParams) => string;
+};
+
+export function useHMACSignatureValidation(
+  options: HMACUpstreamSignatureValidationOptions,
+): YogaPlugin {
+  const extensionName = options.extensionName || DEFAULT_EXTENSION_NAME;
+  let key$: Promise<CryptoKey>;
+  let textEncoder: TextEncoder;
+  const paramsSerializer = options.serializeParams || defaultParamsSerializer;
+
+  return {
+    onParams({ params, fetchAPI }) {
+      textEncoder ||= new fetchAPI.TextEncoder();
+      const extension = params.extensions?.[extensionName];
+      if (!extension) {
+        throw new Error(`Missing HMAC signature: extension ${extensionName} not found in request.`);
```

We throw a raw error ? This means it will be masked, this is what we want ? It will probably make debugging a bit difficult on setup in case something went wrong somewhere

---

It should hard fail in this case, if an unsigned request got to the subgraph, we want to make sure to fail it without details: it might be someone who got access to the internal network. We don't want to hint too much in this case.

---

Ok makes sense :-)

---

Also keep in mind, this error message is logged like any other masked errors.